zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/AzureNotebookRepo.java

```diff
@@ -0,0 +1,201 @@
+        share.getRootDirectoryReference() :
+        share.getRootDirectoryReference().getDirectoryReference(user);
+    userDir.createIfNotExists();
+
+    rootDir = userDir.getDirectoryReference("notebook");
+    rootDir.createIfNotExists();
+  }
+
+  @Override
+  public List<NoteInfo> list() throws IOException {
+    List<NoteInfo> infos = new LinkedList<NoteInfo>();
+    NoteInfo info = null;
+
+    for (ListFileItem item : rootDir.listFilesAndDirectories()) {
+      if (item.getClass() == CloudFileDirectory.class) {
+        CloudFileDirectory dir = (CloudFileDirectory) item;
+
+        try {
+          if (dir.getFileReference("note.json").exists()) {
+            info = new NoteInfo(getNote(dir.getName()));
+
+            if (info != null) {
+              infos.add(info);
+            }
+          }
+        } catch (URISyntaxException e) {
+          LOG.error("Error enumerating notebooks", e);
+        } catch (StorageException e) {
+          LOG.error("Error enumerating notebooks", e);
+        }
+      }
+    }
+
+    return infos;
+  }
+
+  private Note getNote(String noteId) throws IOException {
+    InputStream ins = null;
+
+    try {
+      CloudFileDirectory dir = rootDir.getDirectoryReference(noteId);
+      CloudFile file = dir.getFileReference("note.json");
+
+      ins = file.openRead();
+    } catch (URISyntaxException e) {
+      LOG.error("Error reading file", e);
+      return null;
+    } catch (StorageException e) {
+      LOG.error("Error reading file", e);
+      return null;
+    }
+
+    String json = IOUtils.toString(ins,
+        conf.getString(ZeppelinConfiguration.ConfVars.ZEPPELIN_ENCODING));
+    ins.close();
+
+    GsonBuilder gsonBuilder = new GsonBuilder();
+    gsonBuilder.setPrettyPrinting();
+    Gson gson = gsonBuilder.create();
+
+    Note note = gson.fromJson(json, Note.class);
+
+    for (Paragraph p : note.getParagraphs()) {
+      if (p.getStatus() == Job.Status.PENDING || p.getStatus() == Job.Status.RUNNING) {
+        p.setStatus(Job.Status.ABORT);
+      }
+    }
+
+    return note;
+  }
+
+  @Override
+  public Note get(String noteId) throws IOException {
+    return getNote(noteId);
+  }
+
+  @Override
+  public void save(Note note) throws IOException {
+    GsonBuilder gsonBuilder = new GsonBuilder();
+    gsonBuilder.setPrettyPrinting();
+    Gson gson = gsonBuilder.create();
+    String json = gson.toJson(note);
+
+    ByteArrayOutputStream output = new ByteArrayOutputStream();
+    Writer writer = new OutputStreamWriter(output);
+    writer.write(json);
+    writer.close();
+    output.close();
+
+    byte[] buffer = output.toByteArray();
+
+    try {
+      CloudFileDirectory dir = rootDir.getDirectoryReference(note.getId());
+      dir.createIfNotExists();
+
+      CloudFile cloudFile = dir.getFileReference("note.json");
+      cloudFile.uploadFromByteArray(buffer, 0, buffer.length);
+    } catch (URISyntaxException e) {
+      LOG.error("Error saving file", e);
```

Comment:
got it. IMO that is reasonable since AzureNotebookRepo is acting as a sort of IO/Storage service. An exception from the underlining platform (like Microsoft.WindowsAzure.Storage.StorageException) would need to be wrapped to be communicated back to the caller. I think that is still better than silently swallowing the exception in which case the user would need to ask admin to check the zeppelin log file.

That said, I'll let others chime in.